Dealing with var, parts 1 and 2 (#1696)
dotnet/csharpstandard: standard/patterns.md

The order of evaluation of operations and side effects during pattern-matching (calls to `Deconstruct`, property accesses, and invocations of methods in `System.ITuple`) is not specified.

> The *identifier* `var` appearing in any *pattern* is bound using the normal identifier-resolution rules. The *identifier* `var` may appear in patterns only where the entity to which it resolves is valid in that position. When no declaration of the *identifier* `var` is in scope, the form `var` *designation* is recognised as a *var_pattern* ([§11.2.4](patterns.md#1124-var-pattern)); when any declaration of the *identifier* `var` is in scope, a *var_pattern* cannot be used. The verbatim identifier `@var` and Unicode-escaped equivalents are distinct identifiers from the contextual keyword `var` and are bound and used in patterns according to the entities they denote, like any other identifier.

Reviewer: Not sure about the verbatim identifier part - that would suggest this should compile:

```
using System;

class @var {}

class Test
{
    static void Main()
    {
        object x = "";
        if (x is var y)
        {
            Console.WriteLine("yes");
        }
    }
}
```

... but it doesn't. I *think* you're trying to say that using `@var` in a *pattern* isn't equivalent to using `var`, but if that's the case, we should be more explicit about it.

> *Note*: When a declaration of the *identifier* `var` is in scope, the consequences of the rule above are:
>

> > - If `var` denotes a type, then `var` may be the *type* of a *declaration_pattern* ([§11.2.2](patterns.md#1122-declaration-pattern)).

Reviewer 2: I believe this is incorrect, once `var` is declared as a type then it ***cannot*** be used as the type in a *declaration_pattern* – you must go ugly (`@var` *et al*) – but it can be used in `is *type*`[*].

If `var` is declared as a constant then both `is var` – *constant_pattern* and `is var y` – *var_pattern* are valid. However `is @var y` is invalid as that references the constant…

Or something like that!

[*] This is testing for v8, for v9 `is *type*` vs `is *pattern*` where *pattern* is *type_pattern* throws a spanner, and the rest of the toolbox, into the works as both in theory exist and have different semantics yet are syntactically indistinguishable – at least that’s what the source material says…

> - If `var` denotes a constant, then `var` may be the *constant_expression* of a *constant_pattern* ([§11.2.3](patterns.md#1123-constant-pattern)).

> > - Otherwise (for example, when `var` denotes a field, property, local variable, parameter, or any other entity that is neither a type nor a constant), `var` is not valid in pattern position.

Reviewer: I'm not *sure* what "in pattern position" means here. (I suspect just adding a couple of words will be fine.)

Reviewer: Whatever we decide to do, we should apply to line 191 as well.

Reviewer 2: @jskeet – I think the intention is “is not valid as the first token of a *pattern*”

>
> *end note*

> ### 11.2.3 Constant pattern

Reviewer: This now comes *before* 11.2.2 - I suspect some git operation hasn't quite worked

Reviewer 2: @jskeet – the move is intentional, looks like the renumbering tool hasn’t run yet?

The order of evaluation of operations and side effects during pattern-matching (calls to `Deconstruct`, property accesses, and invocations of members of `System.Runtime.CompilerServices.ITuple`) is not specified.

> The token `var` in a *pattern* is recognised as the contextual keyword of a *var_pattern* ([§11.2.4](patterns.md#1124-var-pattern)) only in the form `var` *designation*. The verbatim identifier `@var` and Unicode-escaped spellings of `var` are never recognised as the contextual keyword of a *var_pattern*; they are ordinary identifiers denoting whatever entity is in scope. A pattern that uses such a spelling is therefore interpreted as a *declaration_pattern*, *constant_pattern*, or other pattern form according to the entity that the identifier denotes.
> 
> When a declaration of the *identifier* `var` is in scope, the following rules apply:
> 
> - If `var` denotes a type, then the plain token `var` cannot be used as the *type* of a *declaration_pattern*; that type shall be named another way, such as by the verbatim identifier `@var`.
> - If `var` denotes a constant, then `var` may be the *constant_expression* of a *constant_pattern* ([§11.2.3](patterns.md#1123-constant-pattern)).
> - Otherwise (for example, when `var` denotes a namespace, field, property, local variable, parameter, or any other entity that is neither a type nor a constant), `var` is not valid as the first token of a *pattern* unless it is used in the form `var` *designation*.
> 

Reviewer 2: This reads, along with references in the changes below back to here stating  (***emphasis added***):

> the general rule for the *identifier* `var` ***in patterns*** is specified in §11.2.1

as though the recognition of `var` in this context is somehow special.

However `var` is recognised according to the token rules of §6.4 and the pattern grammar rules; just as uses of contextual keywords elsewhere are. What may differ in any location where a contextual keyword is used are the disambiguation rules that apply if there is a definition in scope with the same name (§6.4.4).

Those disambiguation rules are usually stated close to the source of the ambiguity, which in this case would be in the subclauses for constant, declaration and var patterns; in this case they have been split between here and the relevant subclauses. I don’t see that this gains anything.

Rewording to address these issues could be considered, e.g. by including references back to §6.4 and making it clear the process follows the usual model.

However instead I suggest simply deleting the text here and handling `var` on a case-by-case basis in the following clauses.

```suggestion
```

> <!-- C# 9 update marker: the committee's empirical testing shows that in C# 8, when `var` names a type, `o is (var)` and `o is (@var)` fail CS8400 because type patterns are not available; in C# 9 they bind as *type_pattern*s and are valid. Draft-v9 should update this boundary for type patterns and parenthesized patterns, while preserving that bare `is T` and pattern `is (T)` are distinct grammar/binding paths. -->
> 

Reviewer 2: I don't think stating that the definition of C# is determined by “empirical testing” is how formal language semantics are decided – even if replacing formal definition with empirical observation might be becoming de rigueur for AI software specification 😉

So I will again delete this text as the simplest option, if is not then at minimum it needs rephrasing.

```suggestion
```

### 11.2.3 Constant pattern

A *constant_pattern* is used to test the value of a pattern input value ([§11.1](patterns.md#111-general)) against the given constant value.

```ANTLR
constant_pattern
    : constant_expression
    ;
```

> 
> When a declaration of the *identifier* `var` in scope resolves to a constant, `var` may appear as the *constant_expression* of a *constant_pattern*; the general rule for the *identifier* `var` in patterns is specified in [§11.2.1](patterns.md#1121-general).

Reviewer 2: The introduction of contextual keywords and discards to C# while also maintaining the ability to use the same identifiers in user-defined ways allows for potential confusion but is generally not remarked upon unless rules, such as disambiguation, are required. So why remark on it here?

In this particular situation if `var` is defined as a constant then the input `var` can itself be a pattern which might to some be potentially confusing with the `var x` pattern… Is this more confusing than the general cases caused by contextual keywords and so worthy of note?

If it is thought not then this addition becomes redundant:

```suggestion
```

If it is deemed worthy of explanation then this addition should become a *Note* explaining the potential confusion:

```suggestion

> *Note*: Come up with a short clear wording stating that using `var` as a constant in a context where patterns are used could be confusing – this is left as an exercise… *end note*
```

Take your pick (or neither of course!)

> *Note*: ANTLR makes the specified choice automatically due to the ordering of the alternatives of *simple_designation*. *end note*

> When a declaration of the *identifier* `var` in scope resolves to a type, the plain token `var` cannot appear as the *type* of a *declaration_pattern*; that type shall be named another way, such as by the verbatim identifier `@var`. The general rule for the *identifier* `var` in patterns is specified in [§11.2.1](patterns.md#1121-general).

Reviewer 2: This is a curious case, why does the declaration pattern differ in its handling of `var` compared to the declaration statement? I.e. if `var` is, say, a user-defined class type; then `var x` declares `x` to be of type `var` when the tokens appear in the context of a declaration statement, and the token sequence is invalid if they appear in the context of a declaration pattern – why the difference?

***Assuming*** this difference is intentional we fortunately we are not required to provide the rationale, we just need to state the rule:

```suggestion
A *declaration_pattern* cannot be used to test that a value has a type named `var` unless that type is referenced using an identifier containing a unicode character escape sequence ([§6.4.2](lexical-structure.md#642-unicode-character-escape-sequences)) or represented by an *Escaped_Identifier* ([§6.4.3](lexical-structure.md#643-identifiers)).
```

An example could be included which, say, declares a `class var` and then uses `@var` and/or `v\u0061r` in declaration patterns. Here is one which is not included in the above suggestion as it feels far too long, I expect others will do better:

> *Example*: Given the definition of a class `var` and a variable `x` of that type:
> ```csharp
> class var { ... }
> ```
>
> Then testing whether a variable `x` is of type `var` using the is-type operator ([§12.14.12.1](expressions.md#1214121-the-is-type-operator)):
> ```csharp
> if (x is var) ...
> ```
> is valid. However a similar test using the is-pattern operator ([§12.14.12.2](expressions.md#1214122-the-is-pattern-operator)):
> ```
> if (x is var y) ...
> ```
> is a compile-time error. However the following:
> ```csharp
> if (x is v\u0061r y) ...
> if (x is @var z) ...
> ```
> are both valid uses of the is-pattern operator with a *declaration_pattern*.
>
> *end example*

> The *type* of a *declaration_pattern* cannot be `dynamic`, because the runtime type test is defined in terms of the is-type operator ([§12.14.12.1](expressions.md#1214121-the-is-type-operator)), which does not permit `dynamic`.

Reviewer 2: The “because…” is surely wrong as nothing has the runtime type of `dynamic`? I suggest just:

```suggestion
The *type* of a *declaration_pattern* cannot be `dynamic`.
```

and avoid getting into why.

>
> The condition of the `if` statement is `true` at runtime and the variable `v` holds the value `3` of type `int` inside the block. After the block the variable `v` is in scope, but not definitely assigned. *end example*

### 11.2.4 Var pattern

A *var_pattern* *matches* every value. That is, a pattern-matching operation with a *var_pattern* always succeeds.

A *var_pattern* is *applicable to* every type.

> A *var_pattern* cannot be used when the plain token `var` would refer to an in-scope type named `var`. See [§11.2.1](patterns.md#1121-general) for the interpretation of `var` when it is the first token of a *pattern* in that case.

Reviewer 2: Following the previous suggested changes we need to delete the x-ref as it now goes nowhere, and a slight rewording suggestion:

```suggestion
A *var_pattern* cannot be used when there is an in-scope type named `var`.
```

Is it worth adding a *Note* that there is/is not no easy workaround in such as case?

    ;
```

> Given a match of an input value to the pattern *type* `(` *subpatterns* `)`, a method is selected by searching in *type* for accessible declarations of `Deconstruct` and selecting one among them using the same rules as for the deconstruction declaration.

Reviewer 2: As far as I can see there is no such thing as a “`deconstruction declaration`”, also found on line 261. Are these two uses accidental (untranslated?) copies from some source material or am I missing something?

> It is an error if a *positional_pattern* omits the type, has a single *subpattern* without an *identifier*, has no *property_subpattern* and has no *simple_designation*. This disambiguates between a *constant_pattern* that is parenthesized and a *positional_pattern*.

Reviewer 2: This is just a repeat of line 33, repeating it is harmless but it should be a true repeat:

```suggestion
If the input can be syntactically recognised as both a *constant_pattern* and a *positional_pattern* then the *constant_pattern* shall be chosen.
```

Note that neither location contains an ANTLR comment as some other similar cases do, in this case it would be:

> *Note*: ANTLR makes the specified choice automatically due to the ordering of the alternatives of *pattern*. *end note*

Adding it to line 33 could be done, I wouldn't add it at both locations.
